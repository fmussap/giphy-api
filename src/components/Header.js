import React from 'react'
import { Link } from 'react-router-dom'
import { connect } from 'react-redux'

export const Header = ({ startLogout }) => (
  <header className='header'>
    <div className='content-container'>
      <div className='header__content'>
        <Link className='header__title' to='/'>
          <h1>Animated GIF Search</h1>
          <img src='/images/giphy-logo.gif' />
        </Link>
      </div>
    </div>
  </header>
)

const mapDispatchToProps = (dispatch) => ({
})

export default connect(null, mapDispatchToProps)(Header)
