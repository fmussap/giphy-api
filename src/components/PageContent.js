import React, { Component } from 'react'

class PageContent extends Component {
  constructor (props) {
    super()
    this.state = {
      src: props.urlStill,
      hover: ''
    }

    this.handleClick = () => {
      props.handleOpenModal(props.urlModal, props.title)
      // this changes the class to disable the hover efect
      this.setState(() => ({
        hover: 'no-hover'
      }))
    }

    // this changes the image to be load from static / animated
    this.handleHoverOver = () => {
      this.setState(() => ({
        src: props.src
      }))
    }

    // it changes the image to be load from animated / static
    this.handleHoverOut = () => {
      this.setState(() => ({
        src: props.urlStill,
        hover: ''
      }))
    }
  }

  render () {
    return (
      <div
        className={`individual-gif-container ${this.state.hover}`}
        style={{ backgroundImage: `url(${this.state.src})`, backgroundRepeat: 'no-repeat' }}
        onClick={this.handleClick}
        onMouseOver={this.handleHoverOver}
        onMouseOut={this.handleHoverOut}
      >
        <div className='individual-gif-info'>{this.props.title}</div>
      </div>
    )
  }
}

export default PageContent
