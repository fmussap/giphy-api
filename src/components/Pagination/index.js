import React from 'react'
import PropTypes from 'prop-types'
import pagination from '../../utils/pagination'
import Page from './page'

const Pagination = ({ total, activePage, pageLink, onClick, max }) => {
  const apage = activePage === 1 ? activePage : ((activePage - 1) / max) + 1
  const prev = '<'
  const next = '>'
  const nextPage = apage + 1 <= 199 ? apage + 1 : 199
  const prevPage = apage - 1 >= 1 ? apage - 1 : 1
  // the giphy development api key only allows to load 199 pages.
  // this projects is using the same key for production and development
  // for production is necessary to ask for a new type of key.
  const limit = process.env.NODE_ENV === 'development' ? 199 : null

  return (
    <ul className='pagination'>
      <Page className='prev' page={prev} pageLink={pageLink.replace('%page%', prev)} onClick={() => onClick(prevPage)} />
      {pagination({ total: total >= 200 ? (!limit ? total : limit) : total, activePage: apage }).map((page, index) => {
        return (
          <div key={index} className={`pagination-item ${apage === page ? 'active' : ''}`}>
            <Page page={page} pageLink={pageLink.replace('%page%', page)} onClick={onClick} />
          </div>
        )
      })}
      <Page className='next' page={next} pageLink={pageLink.replace('%page%', next)} onClick={() => onClick(nextPage)} />
    </ul>
  )
}

Pagination.defaultProps = {
  pageLink: '',
  activePage: 1
}

Pagination.propTypes = {
  total: PropTypes.number,
  activePage: PropTypes.number,
  pageLink: PropTypes.string,
  onClick: PropTypes.func
}

export default Pagination
