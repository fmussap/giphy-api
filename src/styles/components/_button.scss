.button {
  background: $button-bg;
  border: none;
  border-radius: 10px;
  box-shadow: 2px 4px 2px -2px $dark-grey;
  color: $button-color;
  cursor: pointer;
  display: inline-block;
  font-size: $base-font-size;
  line-height: 1.2;
  margin: $s-size;
  outline: none;
  padding: $s-size;
  text-decoration: none;
  text-transform: uppercase;
  transition: background-color .3s ease;
}

.button:hover {
  background-color: $button-hv;
}
