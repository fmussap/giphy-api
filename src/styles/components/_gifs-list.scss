.gifs-list-container {
  align-items: center;
  background: black;
  display: flex;
  flex-wrap: wrap;
  justify-content: center;
}

.gifs-filter-container {
  display: flex;
  justify-content: center;
  margin-top: $s-size;
  width: 100%;
}

.individual-gif {
  cursor: pointer;
  height: 18rem;
  object-fit: cover;
  padding: $s-size;
  width: 18rem;
}

.individual-gif-container {
  height: 15rem;
  margin: 1rem;
  width: 15rem;
  background-size: cover;
  transform: scale(1, 1);
  transition: transform .6s;
  @media (min-width: $desktop-breakpoint) {
    height: 18rem;
    margin: 1.1rem;
    width: 18rem;
  }
}

.individual-gif-container:not(.no-hover):hover {
  object-fit: cover;
  background-size: auto;
  @media (min-width: $desktop-breakpoint) {
    z-index: 99 !important;
    transform: scale(1.5, 1.5);
  }
  .individual-gif-info {
    opacity: 1;
  }
}

.individual-gif-info {
  background: rgba(0, 0, 0, 0.5);
  color: white;
  font-size: 1rem;
  padding: 0.5rem;
  opacity: 0;
  transition:opacity 1s ease-in-out;

}

.modal-container {
  height: 80vw;
  background-size: cover;
  @media (min-width: $desktop-breakpoint) {
    height: 50vw;
  }
}

.search-container {
  display: flex;
  justify-content: center;
  align-items: center;
  margin: $s-size;
  @media (min-width: $desktop-breakpoint) {
    margin: $m-size;
  }
}

.img-search {
  cursor: pointer;
  margin-left: $s-size;
  width: 4rem;
  height: 4rem;
  @media (min-width: $desktop-breakpoint) {
    width: 5rem;
    height: 5rem;
  }
}