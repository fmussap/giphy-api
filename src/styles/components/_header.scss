.header {
  background: black;
  border-style: solid;
  border-left: 0;
  border-right: 0;
  border-top: 0;
  border-bottom-width: .5rem;
  border-bottom-color: white;
  animation-name: fade;
  animation-duration: 3s;
}

.header__content {
  align-items: center;
  display: flex;
  justify-content: space-between;
  padding: $s-size 0;
}

.header__title {
  color: white;
  text-decoration: none;
  h1 {
    margin: 0;
  }
}

@keyframes fade {
  from {opacity: 0.4;}
  to {opacity: 1;}
}
