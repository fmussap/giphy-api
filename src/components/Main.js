import React from 'react'

import Header from './Header'
import Dashboard from './DashboardPage'

export const Main = () => (
  <div className='main'>
    <Header />
    <Dashboard />
  </div>
)

export default Main
