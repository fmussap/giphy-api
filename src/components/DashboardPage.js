import React, { Component } from 'react'
import { connect } from 'react-redux'
import Modal from 'react-modal'
import { CopyToClipboard } from 'react-copy-to-clipboard'

import { getAllGifs, getGifsPage } from '../actions'
import PageContent from './PageContent'
import Pagination from './Pagination'

export class DashboardPage extends Component {
  state = {
    type: null,
    max: 25,
    src: '',
    isOpen: false,
    title: '',
    copyLink: false,
    name: ''
  }

  // render all the GIFs
  renderGifs () {
    if (this.props.gifs && this.props.gifs.length > 0) {
      const gifs = this.props.gifs
      return gifs.map((gif) => {
        const url = gif.images.fixed_width.url
        const urlStill = gif.images.fixed_width_still.url
        const urlModal = gif.images.downsized.url
        const title = gif.title.replace('GIF', '')
        return (
          <PageContent
            handleOpenModal={this.handleOpenModal}
            src={url}
            urlStill={urlStill}
            urlModal={urlModal}
            key={gif.id}
            title={title}
          />
        )
      })
    }
    return <div>No gifs found</div>
  }

  handleOpenModal = (src, title) => {
    this.setState(() => ({ 
      isOpen: true,
      src,
      title,
    }))
  }

  handleCloseModal = () => {
    this.setState(() => ({ 
      isOpen: false,
      src: '',
      title: '',
      copyLink: false
    }))
  }

  handleCopyLink = () => {
    this.setState(() => ({ 
      copyLink: true
    }))
  }

  handleClickAll = (page = 1) => () => {
    this.props.getAllGifs(page, this.state.max)
    this.setState(() => ({ type: null }))
  }

  handleFilterGifs = (type, page = 1) => () => {
    this.setState(() => ({ type }))
    this.props.getGifsPage(type, page)
  }

  // Load GIFs from each new page
  handlePagination = (page=1) => {
    const type = this.state.type
    if (type) {
      this.props.getGifsPage(type, page, this.state.max)
    } else {
      this.props.getAllGifs(page, this.state.max)
    }
  }

  handleSearchGifs = (e) => {
    const name = e.target.value
    this.setState(() => ({ name }))
  }

  onSearch = () => {
    if (this.state.name) {
      this.handleFilterGifs(this.state.name)()
      this.setState(() => ({ name: '' }))
    }
  }

  renderPagination () {
    const numberOfGifs = this.props.gifs ? this.props.gifs.length : 0
    if (numberOfGifs > 0) {
      return (
        <Pagination
          activePage={this.props.pagination.offset}
          max={this.state.max}
          onClick={this.handlePagination}
          total={this.props.pagination.total_count % this.state.max === 0
            ? this.props.pagination.total_count / this.state.max
            : parseInt(this.props.pagination.total_count / this.state.max) - 1} // total page / GIFs per page
        />
      )
    }
  }

  componentWillMount () {
    // required for Modal
    Modal.setAppElement('body')
  }

  componentDidMount () {
    // Load the first GIFs when the page starts for de first time
    this.props.getAllGifs(1)
  }

  render () {
    return (
      <div className='content-container'>
        <div className='search-container'>
          <input
            className='text-input'
            type='text'
            placeholder='Search GIFs'
            autoFocus
            value={this.state.name}
            onChange={this.handleSearchGifs}
          />
          <img className='img-search' title='Search' src='/images/magnify.png' onClick={this.onSearch} />
        </div>
        <div className='gifs-filter-container'>
          <button className='button all' onClick={this.handleClickAll()}>
            All
          </button>
          <button className='button cats' onClick={this.handleFilterGifs('cat')}>
            Cats
          </button>
          <button className='button dogs' onClick={this.handleFilterGifs('dog')}>
            Dogs
          </button>
        </div>
        <Modal
          isOpen={this.state.isOpen}
          contentLabel={this.state.title}
          shouldReturnFocusAfterClose={false}
          onAfterOpen={this.handleFocus}
          onRequestClose={this.handleCloseModal}
          style={{
            overlay: {
              position: 'fixed',
              backgroundColor: 'rgba(64, 64, 64, 0.75)',
            },
            content: {
              textAlign: 'center',
              border: '1px solid #000', background: '#000',
              overflow: 'none',
              borderRadius: '4px', outline: 'none', padding: '1rem',
              top: '50%', left: '50%', right: 'auto', bottom: 'auto',
              marginRight: '-50%', transform: 'translate(-50%, -50%)',
            }
          }}
        >
          <div>
            <img src={this.state.src} style={{ maxWidth: '90%' }}/>
            <div className='modal-title'>{this.state.title}</div>
            {!this.state.copyLink && (
              <CopyToClipboard text={this.state.src}>
                <div className='copy-link' onClick={this.handleCopyLink}>Copy link</div>
              </CopyToClipboard>
            )}
            {this.state.copyLink && <div className='link-message'>Link copied to clipboard</div>}
            <img title='Close' className='close-modal' src='/images/close-circle-outline.png' onClick={this.handleCloseModal} />
          </div>
        </Modal>
        {this.renderPagination()}
        <div  className='gifs-list-container'>
          {this.renderGifs()}
        </div>
        {this.renderPagination()}
      </div>
    )
  }
}

const mapStateToProps = ({ gifs }) => ({
  gifs: gifs.data || [],
  meta: gifs.meta,
  pagination: gifs.pagination
})

export default connect(mapStateToProps, {
  getAllGifs,
  getGifsPage
})(DashboardPage)
