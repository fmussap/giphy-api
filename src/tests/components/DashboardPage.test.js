import React from 'react'
import { shallow } from 'enzyme'
import { DashboardPage } from '../../components/DashboardPage'

let spy, wrapper
beforeEach(() => {
  spy = jest.fn()
  wrapper = shallow(<DashboardPage
    getAllGifs={spy}
    getGifsPage={spy}
  />)
})

test('should render DashboardPage correctly', () => {
  expect(wrapper).toMatchSnapshot()
})

test('should handle handleClickAll on the All button', () => {
  wrapper.find('button.all').simulate('click')
  expect(spy).toHaveBeenLastCalledWith(1, 25)
})

test('should handle handleFilterGifs on the Cats button', () => {
  wrapper.find('button.cats').simulate('click')
  expect(spy).toHaveBeenLastCalledWith('cat', 1)
})

test('should handle handleFilterGifs on the Dogs button', () => {
  wrapper.find('button.dogs').simulate('click')
  expect(spy).toHaveBeenLastCalledWith('dog', 1)
})

test('handleSearchGifs should set a new search on input change', () => {
  const value = 'New search'
  wrapper.find('input').simulate('change', {
    target: { value }
  })
  expect(wrapper.state('name')).toBe(value)
})

test('should handle handleCopyLink', () => {
  const link = 'http://test/src'
  wrapper.setState({ src: link })
  wrapper.find('div.copy-link').simulate('click')
  expect(wrapper.state().src).toBe(link)
})
