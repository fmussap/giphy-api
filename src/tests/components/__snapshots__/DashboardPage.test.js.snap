// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render DashboardPage correctly 1`] = `
<div
  className="content-container"
>
  <div
    className="search-container"
  >
    <input
      autoFocus={true}
      className="text-input"
      onChange={[Function]}
      placeholder="Search GIFs"
      type="text"
      value=""
    />
    <img
      className="img-search"
      onClick={[Function]}
      src="/images/magnify.png"
      title="Search"
    />
  </div>
  <div
    className="gifs-filter-container"
  >
    <button
      className="button all"
      onClick={[Function]}
    >
      All
    </button>
    <button
      className="button cats"
      onClick={[Function]}
    >
      Cats
    </button>
    <button
      className="button dogs"
      onClick={[Function]}
    >
      Dogs
    </button>
  </div>
  <Modal
    ariaHideApp={true}
    bodyOpenClassName="ReactModal__Body--open"
    closeTimeoutMS={0}
    contentLabel=""
    isOpen={false}
    onRequestClose={[Function]}
    parentSelector={[Function]}
    portalClassName="ReactModalPortal"
    shouldCloseOnEsc={true}
    shouldCloseOnOverlayClick={true}
    shouldFocusAfterRender={true}
    shouldReturnFocusAfterClose={false}
    style={
      Object {
        "content": Object {
          "background": "#000",
          "border": "1px solid #000",
          "borderRadius": "4px",
          "bottom": "auto",
          "left": "50%",
          "marginRight": "-50%",
          "outline": "none",
          "overflow": "none",
          "padding": "1rem",
          "right": "auto",
          "textAlign": "center",
          "top": "50%",
          "transform": "translate(-50%, -50%)",
        },
        "overlay": Object {
          "backgroundColor": "rgba(64, 64, 64, 0.75)",
          "position": "fixed",
        },
      }
    }
  >
    <div>
      <img
        src=""
        style={
          Object {
            "maxWidth": "90%",
          }
        }
      />
      <div
        className="modal-title"
      />
      <CopyToClipboard
        text=""
      >
        <div
          className="copy-link"
          onClick={[Function]}
        >
          Copy link
        </div>
      </CopyToClipboard>
      <img
        className="close-modal"
        onClick={[Function]}
        src="/images/close-circle-outline.png"
        title="Close"
      />
    </div>
  </Modal>
  <div
    className="gifs-list-container"
  >
    <div>
      No gifs found
    </div>
  </div>
</div>
`;
